EmbeddedPkg/Library/VirtualRealTimeClockLib/VirtualRealTimeClockLib.c

```diff
@@ -53,7 +53,7 @@ LibGetTime (
   UINT64      Freq;
   UINT64      Counter;
   UINT64      Remainder;
-  UINTN       EpochSeconds;
+  UINT64      EpochSeconds;

```

Comment:
This change introduces a direct bug on 32-bit architectures, as well as breaking compatibility with existing variables.

What is the error flagged by codeQL that this is supposed to address? The commit message does not contain information.

Reply:
`d:\a\edk2\edk2\EmbeddedPkg\Library\VirtualRealTimeClockLib\VirtualRealTimeClockLib.c(106): '+=': conversion from 'UINT64' to 'UINTN', possible loss of data`

`d:\a\edk2\edk2\EmbeddedPkg\Library\VirtualRealTimeClockLib\VirtualRealTimeClockLib.c(280): '-=': conversion from 'UINT64' to 'UINTN', possible loss of data`

I run CodeQL test on my repo branch.

Reply:
Hi Leif,

I would argue that this code has never been functionally correct.

EpochSeconds is a UINT32 or a UINT64, depending on compilation target.

L"RtcEpochSeconds" only stores it as a UINT32 or a UINT64, depending on compilation target, which is not great. This has the potentional to truncate a value. 


Counter       = GetPerformanceCounter ();  // Returns a UINT64
EpochSeconds += DivU64x64Remainder (Counter, Freq, &Remainder);  // returns UINT64

 UINT64 divided by UINT64, can cause truncation, especially with the different implementations of TimerLib that a platform might choose/create for themselves. 

I do, however, see your point that this will break compatibility with 32-bit architectures since the variable size is increasing. If we look at it as this code is already fragile, and it just takes the correct set of circumstances to expose it, then this might be worth the breaking change to correct the logic.

Reply:
@apop5 

I don't disagree. And I suspect the impact of this would be kind of low anyway.

`Size         = sizeof (UINTN);` needs to change to `sizeof (EpochSeconds)` though. It should have been that to start with tbh.

But I need to convince myself of the failure case. I guess what would actually happen (with the sizeof change) is any existing 32-bit variable would be read into a 64-bit buffer. But all supported architectures are little-endian, so the end result should actually work transparently.

All the SetVariable calls actually get the size correct anyway, but the `EpochSeconds` variable in `LibSetTime()` also needs to be turned into a UINT64.

Reply:
I agree with the low impact, for now. As we get closer to 2038, it might be a bigger problem.

I agree with your additional suggested changes as well.